Add an API Guide for Configuration Variables (#691)
fermyon/developer: content/spin/variables.md

title = "Configuration Variables"
template = "spin_main"
date = "2022-06-14T00:22:56Z"
enable_shortcodes = true
[extra]
url = "https://github.com/fermyon/developer/blob/main//content/spin/variables.md"

---
- [Using the Config SDK to Get Variables](#using-the-config-sdk-to-get-variables)

Spin supports dynamic application variables. Instead of being static, their values can be updated without modifying the application, creating a simpler experience for rotating secrets, updating API endpoints, and more. 

> These variables are defined in a Spin application manifest (in the `[variables]` section) and are provided by a [configuration provider](/spin/dynamic-configuration.md#custom-config-providers). When running Spin locally, the configuration provider can be Vault for secrets or host environment variables. Refer to the [dynamic configuration documentation](/spin/dynamic-configuration.md) to learn how to configure variables locally.

Reviewer: ```suggestion
These variables are defined in a Spin application manifest (in the `[variables]` section), and their values can be set or overridden at runtime by a [configuration provider](/spin/dynamic-configuration.md#custom-config-providers). When running Spin locally, the configuration provider can be Hashicorp Vault for secrets, or host environment variables. Refer to the [dynamic configuration documentation](/spin/dynamic-configuration.md) to learn how to configure variables locally.
```

> ## Using the Config SDK to Get Variables

Reviewer: The rubric in other API guides has been "How to Use X From Applications" - suggest avoiding "config SDK" as if it was as separate SDK.

> The Spin SDK surfaces the Spin Config interface to your language. The [interface](https://github.com/fermyon/spin/blob/main/wit/ephemeral/spin-config.wit) consists of one operation:

Reviewer: ```suggestion
The Spin SDK surfaces the Spin configuration interface to your language. The [interface](https://github.com/fermyon/spin/blob/main/wit/ephemeral/spin-config.wit) consists of one operation:
```

| Operation  | Parameters                          | Returns             | Behavior |
|------------|-------------------------------------|---------------------|----------|
| `get-config`    | Variable name  | Variable value    | Gets the value of the variable from the configured provider |

> In order to get a variable, it must be declared in an application's manifest (`spin.toml`). For example, say an application needs to access a secret. A variable named `password` could be added to the manifest and set as required with `required = true` since there is no reasonable default value for a secret. To do this, add a top-level `[variables]` section to the application manifest (`spin.toml`) and declare the variable within it.

Reviewer: From an application authoring point of view, this feels a little backwards.  In order to _get_ a variable (application concern), it has to be declared on the component.  And then that harks back to an application variable, which is how you _set_ variables (operator concern).

This probably needs more explanation than can fit conveniently between "here is the underlying WIT API" and "here is how it gets surfaced in the SDKs".  Maybe a separate section?

<!-- @selectiveCpy -->
```toml
# Add this above the [component] section
[variables]
password = { required = true }
```

To surface the variable to a component, add a `[component.config]` section in the component and specify the variable within it. Instead of statically assigning the value of the config variable, dynamically reference the variable with [mustache](https://mustache.github.io/)-inspired string templates. Only components that explicitly use the variables in their configuration section will get access to them. This enables only exposing variables and secrets to the desired components of an application.

```toml
# Add this below the [component.build] section
[component.config]

> password = "\{{ password }}"

Reviewer: It would be really good to use an example where application variables and component config values have different names and values - I think this would help illustrate the distinction and the flow.  I am not sure of a good example though - maybe database host and access token as variables, and connection string as the config value?  (Unfortunately I suspect our real connection strings need more inputs than that...)

Author: thats a good example but i think i am already getting a little verbose with the examples here. I could change it to something mocked with a fake db host and password but i am not sure i could add clean working examples

Reviewer: No worries then, ship it as is.  We can always iterate if and when we feel motivated.

Author: I added a static mapping to the "Adding Variables to an Application" section but am not using that config variable in the example implementations

[variables]
password = { required = true }

[[component]]
id = "password_checker"
source = "app.wasm"
[component.trigger]
route = "/..."
[component.build]
command = "spin py2wasm app -o app.wasm"
[component.config]
password = "\{{ password }}"
```

To illustrate the config API, each of the following examples receives a password via the HTTP request body, compares it to an expected password, and returns a JSON response indicating whether the submitted password matched or not. To use the [environment variable provider](/spin/dynamic-configuration.md#environment-variable-provider) to set the variable values locally, set the `password` variable's value in an environment variable prefixed with `SPIN_CONFIG_`. The provider gets the variable values from the `spin` process's environment:

<!-- @selectiveCpy -->
```bash
$ SPIN_CONFIG_PASSWORD="123" spin build --up

Serving http://127.0.0.1:3000
Available Routes:
  password-checker: http://127.0.0.1:3000 (wildcard)
```

Send a request to the application with the correct password in the body to authenticate successfully.

<!-- @selectiveCpy -->
```bash
$ curl -d "123" http://127.0.0.1:3000
{"authentication": "accepted"}
```

> The exact details of calling the config SDK from a Spin application depends on the language:

Reviewer: I do think this is a bit far from the initial "this is the config interface".  Let's break this page into sections, how to use config from an application, and how to author the config stuff in the manifest.  (I dunno which way round they should go though!)

Author: I created a new "Adding Variables to Your Applications" section that just focuses on the manifest. Let me know what you think

Reviewer: Looks good! Thanks!

{{ tabs "sdk-type" }}

{{ startTab "Rust"}}

> The config function is available in the `spin_sdk::config` module.

Reviewer: ```suggestion
The config function is available in the `spin_sdk::config` module and is named `get`.
```

(mentioning this because it is named `get_config` in the WIT mentioned above)

```rust
use anyhow::Result;
use spin_sdk::{
    config,
    http::{Request, Response},
    http_component,
};

#[http_component]
fn handle_spin_example(req: Request) -> Result<Response> {
    let password = std::str::from_utf8(req.body().as_ref().unwrap()).unwrap();
    let expected = config::get("password").expect("could not get variable");
    let response = if expected == password {
        "accepted"
    } else {
        "denied"
    };
    let response_json = format!("\{{\"authentication\": \"{}\"}}", response);
    Ok(http::Response::builder()
        .status(200)
        .header("Content-Type", "application/json")
        .body(Some(response_json.into()))?)
}

```

{{ blockEnd }}

{{ startTab "TypeScript"}}

> The config function is available in the `spinSdk.config` package.

Reviewer: ```suggestion
The config function is available in the `spinSdk.config` package and is named `get`.
```

```ts
import { HandleRequest, HttpRequest, HttpResponse } from "@fermyon/spin-sdk"

const decoder = new TextDecoder("utf-8")

export const handleRequest: HandleRequest = async function (request: HttpRequest): Promise<HttpResponse> {
  const expected = decoder.decode(request.body)
  let password = spinSdk.config.get("openai_key")
  let access = "denied"
  if (expected === password) {
      access = "accepted"
  }
  let responseJson = `{\"authentication\": \"${access}\"}`;

  return {
    status: 200,
    headers: { "Content-Type": "application/json" },
    body: responseJson
  }
}

```

{{ blockEnd }}

{{ startTab "Python"}}

> The config function is available in the `spin_config` package.

Reviewer: ```suggestion
The config function is available in the `spin_config` package and is named `config_get`.
```

```py
from spin_http import Response
from spin_config import config_get

def handle_request(request):
    password = request.body.decode("utf-8")
    expected = config_get("password")
    access = "denied"
    if expected == password:
        access = "accepted"
    response = f'\{{"authentication": "{access}"}}'
    return Response(200,
                    {"content-type": "application/json"},
                    bytes(response, "utf-8"))
```

{{ blockEnd }}

{{ startTab "TinyGo"}}

> The config function is available in the `github.com/fermyon/spin/sdk/go/config` package. See [Go package](https://pkg.go.dev/github.com/fermyon/spin/sdk/go/config) for reference documentation.

Reviewer: ```suggestion
The config function is available in the `github.com/fermyon/spin/sdk/go/config` package and is named `Get`. See [Go package](https://pkg.go.dev/github.com/fermyon/spin/sdk/go/config) for reference documentation.
```
